packages/klighd-core/src/search-bar/searchbar-actions.ts

```diff
@@ -0,0 +1,680 @@
+        return action.kind === KIND
+    }
+}
+
+@injectable()
+export class ToggleSearchBarHandler implements IActionHandler {
+    handle(action: Action): void {
+        if (ToggleSearchBarAction.isThisAction(action)) {
+            if (action.id !== SearchBar.ID) return
+
+            const newVisible = action.state === 'show'
+
+            if (action.panel.isVisible !== newVisible) {
+                action.panel.changeVisibility(newVisible)
+                action.panel.update()
+            }
+        }
+    }
+}
+
+export interface UpdateHighlightsAction extends Action {
+    kind: typeof UpdateHighlightsAction.KIND
+    selectedIndex: number
+    previousIndex: number | undefined
+    results: SearchResult[]
+    panel: SearchBarPanel
+}
+
+// eslint-disable-next-line no-redeclare
+export namespace UpdateHighlightsAction {
+    export const KIND = 'updateHighlights'
+
+    export function create(
+        currentIndex: number,
+        prevIndex: number | undefined,
+        results: SearchResult[],
+        panel: SearchBarPanel
+    ): UpdateHighlightsAction {
+        return {
+            kind: KIND,
+            selectedIndex: currentIndex,
+            previousIndex: prevIndex,
+            results,
+            panel,
+        }
+    }
+
+    export function isThisAction(action: Action): action is UpdateHighlightsAction {
+        return action.kind === KIND
+    }
+}
+
+export interface ClearHighlightsAction extends Action {
+    kind: typeof ClearHighlightsAction.KIND
+}
+
+// eslint-disable-next-line no-redeclare
+export namespace ClearHighlightsAction {
+    export const KIND = 'clearHighlights'
+
+    export function create(): ClearHighlightsAction {
+        return {
+            kind: KIND,
+        }
+    }
+
+    export function isThisAction(action: Action): action is ClearHighlightsAction {
+        return action.kind === KIND
+    }
+}
+
+function createHighlightRectangle(
+    elem: SModelElement,
+    xPos: number,
+    yPos: number,
+    width: number,
+    height: number,
+    highlight: number
+): KRectangle {
+    return {
+        type: 'KRectangleImpl',
+        id: `highlightRect-${elem.id}`,
+        children: [],
+        properties: {
+            x: xPos,
+            y: yPos,
+            width,
+            height,
+            'klighd.rendering.highlight': highlight,
+        },
+        actions: [],
+        styles: [],
+    }
+}
+
+/**
+ * Remove all highlights
+ * @param elem the root
+ */
+function removeHighlights(elem: SModelElement): void {
+    const queue: SModelElement[] = [elem]
+
+    while (queue.length > 0) {
+        const element = queue.shift()!
+
+        // Remove KText highlights
+        if (isKText(element) && element.styles) {
+            element.properties['klighd.rendering.highlight'] = 0
+        }
+
+        // Remove rectangle highlights from children
+        if ('children' in element && Array.isArray(element.children)) {
```

Comment:
Is this a hacky way to loop through graph elements and renderings, as they both have a "children" element? If so, this seems dangerous to do the same type of loop for both types of elements.